Add support for userData in Azure

bpradipt opened the pull request:
Fix #1315

mkulke reviewed (commented):
I glanced over it and will look at it more deeply asap. Some thoughts so far:

- Are we sure want to support both cloud-init based initialization and homegrown user-data parsing? What would be the upside.
- Do we need to have an option for IMDS timeout or would a sensible default be enough?
- There is some mention of CustomData, afaik CustomData has been superseded by UserData and we probably should not use it in new code.

bpradipt commented on cmd/agent-protocol-forwarder/main.go:
```diff
@@ -29,6 +42,207 @@ type Config struct {
+	// Split the agentConfigFile into lines
+	lines := strings.Split(string(agentConfig), "\n")
+
+	// Loop through the lines and replace the line that starts with aa_kbc_params
+	for i, line := range lines {
+		if strings.Contains(line, "aa_kbc_params") {
+			lines[i] = fmt.Sprintf("aa_kbc_params = \"%s\"", aaKBCParams)
+			fmt.Printf("Updated line: %s\n", lines[i])
+		}
+	}
```
good suggestion. I'll test it in the next iteration

bpradipt replied:
I have not included a toml parser yet since realised that changing the entire line with `aa_kbc_params` should suffice for now. 
Let me know if it's fine to use a toml parser as a future improvement?

bpradipt commented on cmd/agent-protocol-forwarder/main.go:
```diff
@@ -75,6 +299,64 @@ func (cfg *Config) Setup() (cmd.Starter, error) {
 
+	if enableUserData {
+		// Use retry.Do to retry the getUserData function until it succeeds
+		// This is needed because the VM's userData is not available immediately
```
In my tests, the IMDS endpoint was available but it took some time for the userData to be available.

mkulke replied:
hmm... that's not great, if we have to wait a couple of seconds for the data to be available that offsets what we removed by not relying on cloud-init (or maybe that's why cloud-init is slow to begin with).

bpradipt replied:
I kept a default of 3 min (higher side) which can be changed. We can revisit the timeout later on.

bpradipt commented on cmd/agent-protocol-forwarder/main.go:
```diff
@@ -29,6 +42,203 @@ type Config struct {
+	decoded, err := base64.StdEncoding.DecodeString(string(body))
+	if err != nil {
+		fmt.Printf("failed to decode b64 encoded userData: %s", err)
+		return ""
```
I'm bailing out later in the retry.Do loop if userData is empty. As a matter of convenience I just preferred to handle a single return instead of multiple return values.  Do you suggest to bail out here ? I can rewrite the method.

mkulke replied:
hmm, I think it makes sense to make this function fail with an error locally, not on the caller site. Later things might get refactored and the empty-string == error assumption could get lost.

mkulke commented on pkg/adaptor/cloud/azure/provider.go:
```diff
@@ -450,3 +387,108 @@ func (p *azureProvider) updateInstanceSizeSpecList() error {
+		Tags: tags,
+	}
+
+	// If EnableUserData is set to false then set OSProfile.CustomData to b64EncData and
+	// armcompute.VirtualMachine.Properties.UserData to nil
+	// If EnableUserData is set to true then set armcompute.VirtualMachine.Properties.UserData to b64EncData and
+	// OSProfile.CustomData to nil
```
I wouldn't differentiate between CustomData + UserData. As I understand it, CustomData is merely an older way to pass-in user data via a UDF filesystem to a VM. So, what we are actually doing is writing the file in Cloud-Config format for cloud-init or in our own specific format (as raw json for example). 

Whether user-data/custom-data is picked and parsed by cloud-init depends on the first line:

![image](https://github.com/confidential-containers/cloud-api-adaptor/assets/273280/520667a8-e8ef-48af-926f-5d5b6bf61729)

[see here](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/using-cloud-init#deploying-a-cloud-init-enabled-virtual-machine)

Hence I would suggest to always in UserData

mkulke commented:
> Isn't `agent-protocol-forwarder` the wrong place to change the `agent-config.toml`. AFAIU it's parsed by kata-agent at startup, and changes won't get picked up.

@bpradipt ☝️ this is something I don't understand yet. Can you clarify.

What I had in mind was a one shot process at that modifies the string before kata-agent and agent-protocol-forwarder start up. E.g. a few lines of shell that put the daemon.json file where it belongs and change the aakbcparams string. Admittedly doing it in golang is a bit more robust, but maybe we could have a seperate `cmd/process-user-data` for that?

mkulke commented on aws/image/centos/aws-centos.pkr.hcl:
```diff
@@ -85,6 +85,7 @@ build {
     environment_vars = [
         "CLOUD_PROVIDER=${var.cloud_provider}",
         "PODVM_DISTRO=${var.podvm_distro}",
+        "ENABLE_USERDATA=${var.enable_userdata}"
```
nevermind, I saw below that we indeed toggle between user-data and custom-data

bpradipt commented on pkg/adaptor/cloud/azure/provider.go:
```diff
@@ -149,15 +149,38 @@ func (p *azureProvider) createNetworkInterface(ctx context.Context, nicName stri
+				      {
+				       ...
+					  }
+				  - path: /etc/attestation-agent/kbc-params
```
This was a typo. It's not needed as I'm adding it to the daemon.Config struct.

bpradipt commented:
> > Isn't `agent-protocol-forwarder` the wrong place to change the `agent-config.toml`. AFAIU it's parsed by kata-agent at startup, and changes won't get picked up.
> 
> @bpradipt ☝️ this is something I don't understand yet. Can you clarify.
> 
> What I had in mind was a one shot process at that modifies the string before kata-agent and agent-protocol-forwarder start up. E.g. a few lines of shell that put the daemon.json file where it belongs and change the aakbcparams string. Admittedly doing it in golang is a bit more robust, but maybe we could have a seperate `cmd/process-user-data` for that?

Right. I was thinking kata-agent will reload it the file gets changed :-(..  Will explore adding a cmd/process-user-data.

bpradipt commented:
@mkulke if you get time, can you please try the new changes ?
For building an image you'll need to run this command
`LIBC=gnu CLOUD_PROVIDER=azure PODVM_DISTRO=ubuntu ENABLE_USERDATA=1 make image`

And then for CAA, you'll need to include the following additional prams in the configmap
```
ENABLE_USERDATA: "true"
AA_KBC_PARAMS: "cc_kbc::http://192.168.100.2:8080"
```

bpradipt commented:
Another aspect which I'm not yet sure how to make the process-user-data service optional for cases where config-drive is used.

bpradipt commented:
@mkulke I think I have addressed all the comments in this iteration.
For trying out, please build the image using the following cli
```
LIBC=gnu CLOUD_PROVIDER=azure PODVM_DISTRO=ubuntu DISABLE_CLOUD_CONFIG=true make image
```

Make the following changes in the configmap
```
DISABLE_CLOUD_CONFIG: "true"
AA_KBC_PARAMS: "cc_kbc::http://192.168.100.2:8080"
```

I am yet to test with real KBS. Just tested spinning up a pod using both the approaches - fetching userdata from cloud-config (current method) and fetching userdata from metadata service

mkulke commented:
> @mkulke I think I have addressed all the comments in this iteration. For trying out, please build the image using the following cli
> 
> ```
> LIBC=gnu CLOUD_PROVIDER=azure PODVM_DISTRO=ubuntu DISABLE_CLOUD_CONFIG=true make image
> ```
> 
> Make the following changes in the configmap
> 
> ```
> DISABLE_CLOUD_CONFIG: "true"
> AA_KBC_PARAMS: "cc_kbc::http://192.168.100.2:8080"
> ```
> 
> I am yet to test with real KBS. Just tested spinning up a pod using both the approaches - fetching userdata from cloud-config (current method) and fetching userdata from metadata service

I'll test it w/ a kbs deployment

mkulke commented:
> Another aspect which I'm not yet sure how to make the process-user-data service optional for cases where config-drive is used.

This is at build time or at runtime? I think we should be able to include/symlink the relevant files at build time. We can probably use systemd overrides for the kata-agent unit to include the additional dependency on the process-user-data unit:

https://www.flatcar.org/docs/latest/setup/systemd/drop-in-units/

bpradipt commented:
I'm hitting the following error when using the same PR changes on AWS :-( 
```
systemd[1]: multi-user.target: Found ordering cycle on kata-agent.service/start
systemd[1]: multi-user.target: Found dependency on process-user-data.service/start
systemd[1]: multi-user.target: Found dependency on cloud-init.target/start
systemd[1]: multi-user.target: Found dependency on multi-user.target/start
systemd[1]: multi-user.target: Job kata-agent.service/start deleted to break ordering cycle starting with multi-user.target/start
systemd[1]: kata-agent.service: Found ordering cycle on process-user-data.service/start
systemd[1]: kata-agent.service: Found dependency on cloud-init.target/start
systemd[1]: kata-agent.service: Found dependency on multi-user.target/start
systemd[1]: kata-agent.service: Found dependency on kata-agent.service/start
systemd[1]: kata-agent.service: Job kata-agent.service/start deleted to break ordering cycle starting with kata-agent.service/start
```

bpradipt commented:
> > Another aspect which I'm not yet sure how to make the process-user-data service optional for cases where config-drive is used.
> 
> This is at build time or at runtime? I think we should be able to include/symlink the relevant files at build time. We can probably use systemd overrides for the kata-agent unit to include the additional dependency on the process-user-data unit:
> 
> https://www.flatcar.org/docs/latest/setup/systemd/drop-in-units/

Currently, I'm adding the process-user-data service by default and changing the option based on whether cloud-config needs to be enabled or disabled. That seems to avoid using systemd overrides.

mkulke commented on cmd/process-user-data/main.go:
```diff
@@ -0,0 +1,354 @@
+
+	})
+
+	cmd.ShowVersion(programName)
+
+	if showVersion {
+		cmd.Exit(0)
+	}
```
```suggestion
	if showVersion {
	        cmd.ShowVersion(programName)
		cmd.Exit(0)
	}
```

mkulke commented on cmd/process-user-data/main.go:
```diff
@@ -0,0 +1,354 @@
+		cmd.Exit(0)
+	}
+
+	if disableCloudConfig {
```
I assume we have this noop flag here because we cannot tell at runtime whether we need to run he tool? I think we can define that at buildtime (see the comment on systemd overrides)

mkulke commented:
It doesn't work reliable for me. It got stuck at `2023/09/06 08:15:58 [adaptor/proxy] Trying to establish agent proxy connection to 10.200.28.11:15150` in the daemonset logs.

I logged into the podvm to verify, and indeed the unit did not run successfully:

```
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found ordering cycle on cloud-init.target/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on cloud-final.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on multi-user.target/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on attestation-agent.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on kata-agent.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on process-user-data.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Job cloud-init.target/start deleted to break ordering cycle starting with process-user-data.service/start
Sep 06 07:55:34 localhost.localdomain process-user-data[337]: process-user-data version v0.7.0-dev
Sep 06 07:55:34 localhost.localdomain process-user-data[337]:   commit: 7b9136c1e19953b92df96b1b82fd2964a0a9d00c-dirty
Sep 06 07:55:34 localhost.localdomain process-user-data[337]:   go: go1.20.3
Sep 06 07:55:34 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: failed to send request: Get "http://169.254.169.254/metadata/instance/compute?api-version=2021-01-01": dial tcp 169.254.169.254:80: connect:>
Sep 06 07:55:34 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: userDataUrl:
Sep 06 07:55:34 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 0: invalid userdata
Sep 06 07:55:39 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 1: invalid userdata
Sep 06 07:55:44 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 2: invalid userdata
Sep 06 07:55:49 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 3: invalid userdata
Sep 06 07:55:54 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 4: invalid userdata
Sep 06 07:55:59 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 5: invalid userdata
Sep 06 07:56:04 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 6: invalid userdata
Sep 06 07:56:09 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 7: invalid userdata
Sep 06 07:56:14 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 8: invalid userdata
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Retry attempt 9: invalid userdata
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Error: Failed to get valid user data
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Wrote userData to file: /peerpod/daemon.json
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: failed to unmarshal userData: unexpected end of JSON input
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: Error: Failed to get daemon config from userData
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 process-user-data[337]: /usr/local/bin/process-user-data: invalid userdata
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 systemd[1]: process-user-data.service: Main process exited, code=exited, status=1/FAILURE
Sep 06 07:56:19 podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586 systemd[1]: process-user-data.service: Failed w
```

I was able to retrieve the metadata, though:

```bash
curl http://169.254.169.254/metadata/instance/compute?api-version=2021-01-01 -H "Metadata: true"
{"azEnvironment":"AzurePublicCloud","customData":"","evictionPolicy":"","isHostCompatibilityLayerVm":"true","licenseType":"","location":"eastus","name":"podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586","offer":"","osProfile":{"adminUsername":"azureuser","computerName":"podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586","disablePasswordAuthentication":"true"},"osType":"Linux","placementGroupId":"","plan":{"name":"","product":"","publisher":""},"platformFaultDomain":"0","platformUpdateDomain":"0","priority":"","provider":"Microsoft.Compute","publicKeys":[{"keyData":"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC/vCXg2rDv0IigT14gPP+IjLLFiUn2IwLo/qiha9OKHTPazPfIo7Ffq23iNJDohW5abtABYdSMj9C8UM1NCDcQIvD5ueF4BaXAeWsGNch/TyhmFYUn9rYvPO7UHLtIYX90daIjkdkBN6O+TbGV8U/VGLMFOmb/7K6AZvR/YHbI7JrKNKZd7ENTUOFTnMAbfOBismSJZV6YQ1CzAyr8+I3hYIUNVekKo87P8WEva65f4Nv4O12isrTMYFM9iTlOR6VleVMUDqcUFNvWZuZu1AN5A2oMqsMd5LsOErvVekrM7Y+QlUU5b0Ed33dBgK7lrDJAmu8yNCbLlkwIuVvnMwIZ magnuskulke@MININT-QQ6OGSF\n","path":"/home/azureuser/.ssh/authorized_keys"}],"publisher":"","resourceGroupName":"mgns","resourceId":"/subscriptions/REDACTED/resourceGroups/mgns/providers/Microsoft.Compute/virtualMachines/podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586","securityProfile":{"secureBootEnabled":"true","virtualTpmEnabled":"true"},"sku":"","storageProfile":{"dataDisks":[],"imageReference":{"id":"","offer":"","publisher":"","sku":"","version":""},"osDisk":{"caching":"ReadWrite","createOption":"FromImage","diffDiskSettings":{"option":""},"diskSizeGB":"30","encryptionSettings":{"enabled":"false"},"image":{"uri":""},"managedDisk":{"id":"/subscriptions/REDACTED/resourceGroups/mgns/providers/Microsoft.Compute/disks/podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586-disk","storageAccountType":"Standard_LRS"},"name":"podvm-nginx-caa-5c4fc7b6c-xvktd-9e01c586-disk","osType":"Linux","vhd":{"uri":""},"writeAcceleratorEnabled":"false"}},"subscriptionId":"REDACTED","tags":"","tagsList":[],"userData":"ewogICAgICAgICAgInBvZC1uZXR3b3JrIjogewogICAgICAgICAgICAgICJwb2RpcCI6ICIxMC4yNDQuMC4yMTcvMjQiLAogICAgICAgICAgICAgICJwb2QtaHctYWRkciI6ICI5YTo2ZTo1MjozNzoyYzo1OSIsCiAgICAgICAgICAgICAgImludGVyZmFjZSI6ICJldGgwIiwKICAgICAgICAgICAgICAid29ya2VyLW5vZGUtaXAiOiAiMTAuMjAwLjI4LjUvMjQiLAogICAgICAgICAgICAgICJ0dW5uZWwtdHlwZSI6ICJ2eGxhbiIsCiAgICAgICAgICAgICAgInJvdXRlcyI6IFsKICAgICAgICAgICAgICAgICAgewogICAgICAgICAgICAgICAgICAgICAgIkRzdCI6ICIiLAogICAgICAgICAgICAgICAgICAgICAgIkdXIjogIjEwLjI0NC4wLjEiLAogICAgICAgICAgICAgICAgICAgICAgIkRldiI6ICJldGgwIgogICAgICAgICAgICAgICAgICB9CiAgICAgICAgICAgICAgXSwKICAgICAgICAgICAgICAibXR1IjogMTUwMCwKICAgICAgICAgICAgICAiaW5kZXgiOiAwLAogICAgICAgICAgICAgICJ2eGxhbi1wb3J0IjogODQ3MiwKICAgICAgICAgICAgICAidnhsYW4taWQiOiA1NTUwMDAsCiAgICAgICAgICAgICAgImRlZGljYXRlZCI6IGZhbHNlCiAgICAgICAgICB9LAogICAgICAgICAgInBvZC1uYW1lc3BhY2UiOiAiZGVmYXVsdCIsCiAgICAgICAgICAicG9kLW5hbWUiOiAibmdpbngtY2FhLTVjNGZjN2I2Yy14dmt0ZCIsCiAgICAgICAgICAidGxzLXNlcnZlci1rZXkiOiAiLS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tXG5NSUdIQWdFQU1CTUdCeXFHU000OUFnRUdDQ3FHU000OUF3RUhCRzB3YXdJQkFRUWdWMlBoT2FBcGVMT3cvcWN4XG5WUlEyUFhLbm02U1gwMFpFc2pLTm5PR3gzRW1oUkFOQ0FBUlVwbklxNGNBNlE2VGhFVWlsTEJDdUJzR2F6UTdIXG5wQ1BHT3hiRWZLTWdZTmxidmU1MEFvZ2llV25kYms3Y1k4YU1ocmNOeEtqaGt5SXlsdkpIUGtWSVxuLS0tLS1FTkQgUFJJVkFURSBLRVktLS0tLVxuIiwKICAgICAgICAgICJ0bHMtc2VydmVyLWNlcnQiOiAiLS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tXG5NSUlDQlRDQ0FhcWdBd0lCQWdJUU15M29WRFViVHRvekNsZkIzeFIyWGpBS0JnZ3Foa2pPUFFRREFqQWpNU0V3XG5Id1lEVlFRS0V4aGhaMlZ1ZEMxd2NtOTBiMk52YkMxbWIzSjNZWEprWlhJd0hoY05Nak13T1RBMk1EYzBPVFU0XG5XaGNOTWpVd09UQTFNRGMwT1RBM1dqQldNU0V3SHdZRFZRUUtFeGhoWjJWdWRDMXdjbTkwYjJOdmJDMW1iM0ozXG5ZWEprWlhJeE1UQXZCZ05WQkFNVEtIQnZaSFp0TFc1bmFXNTRMV05oWVMwMVl6Um1ZemRpTm1NdGVIWnJkR1F0XG5PV1V3TVdNMU9EWXdXVEFUQmdjcWhrak9QUUlCQmdncWhrak9QUU1CQndOQ0FBUlVwbklxNGNBNlE2VGhFVWlsXG5MQkN1QnNHYXpRN0hwQ1BHT3hiRWZLTWdZTmxidmU1MEFvZ2llV25kYms3Y1k4YU1ocmNOeEtqaGt5SXlsdkpIXG5Qa1ZJbzRHTU1JR0pNQTRHQTFVZER3RUIvd1FFQXdJQ0JEQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNXG5CZ05WSFJNQkFmOEVBakFBTUI4R0ExVWRJd1FZTUJhQUZLa3BQeTErSStvUDlKaUdJK09nVndjNVVub1hNRE1HXG5BMVVkRVFRc01DcUNLSEJ2WkhadExXNW5hVzU0TFdOaFlTMDFZelJtWXpkaU5tTXRlSFpyZEdRdE9XVXdNV00xXG5PRFl3Q2dZSUtvWkl6ajBFQXdJRFNRQXdSZ0loQUoyMURpbG5kcjhuK0VueEx4eXdxMkQvS05KRUIzQ2pLUnZjXG4raENUbVhGQUFpRUE4bDVobzN2R3RhbWtjakhtcG0zbHI5bjZ3eVNDZ2hVa2hQL3IzVGpEZXBvPVxuLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLVxuIiwKICAgICAgICAgICJ0bHMtY2xpZW50LWNhIjogIi0tLS0tQkVHSU4gQ0VSVElGSUNBVEUtLS0tLVxuTUlJQmJUQ0NBUktnQXdJQkFnSVJBSkVSMHRUK1ZnUmRYN3U5MGtVaU85d3dDZ1lJS29aSXpqMEVBd0l3SERFYVxuTUJnR0ExVUVDaE1SWTJ4dmRXUXRZWEJwTFdGa1lYQjBiM0l3SGhjTk1qTXdPVEEyTURjME9UQTNXaGNOTWpVd1xuT1RBMU1EYzBPVEEzV2pBY01Sb3dHQVlEVlFRS0V4RmpiRzkxWkMxaGNHa3RZV1JoY0hSdmNqQlpNQk1HQnlxR1xuU000OUFnRUdDQ3FHU000OUF3RUhBMElBQk1OMWF2Yk15L0tkTTFmVi9JVGVGWmN2QWsyVU9YMWx0V0VmQ3Q0a1xuVHVEaUs5SHc0TDluVWxCRE14b3NMbWEveUVDYVZQNHRLdVgrQm0xb0pLekx3SWFqTlRBek1BNEdBMVVkRHdFQlxuL3dRRUF3SUNCREFUQmdOVkhTVUVEREFLQmdnckJnRUZCUWNEQWpBTUJnTlZIUk1CQWY4RUFqQUFNQW9HQ0NxR1xuU000OUJBTUNBMGtBTUVZQ0lRRFNUUk9jUkFLN3ZSNU1WODRnc08yb2sxU1NHSnQ0SlgvWWllOFNXZkw1cWdJaFxuQVBvNmVSNVIxaVcxUS9DQ05oSTAyZmV2ZVM5OUp4VjMzd0wzenkwMGVpOGFcbi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS1cbiIsCiAgICAgICAgICAiYWEta2JjLXBhcmFtcyI6ICJjY19rYmM6Omh0dHA6Ly8xMC4yMDAuMjguNTozMDA4MCIKICAgICAgfQ==","version":"20230906.0.1","vmId":"6097895b-2920-483d-a154-bd77138ba218","vmScaleSetName":"","vmSize":"Standard_DC2as_v5","zone":""}
```

mkulke commented on podvm/files/etc/systemd/system/process-user-data.service:
```diff
@@ -0,0 +1,17 @@
+
+[Unit]
+Description=Process user data
+After=cloud-init.target
```
Why depend on `cloud-init` here? I think this yields a dependency cycle:

```
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on cloud-final.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on multi-user.target/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on attestation-agent.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on kata-agent.service/start
Sep 06 07:55:32 localhost.localdomain systemd[1]: process-user-data.service: Found dependency on process-user-data.service/start
```

mkulke replied:
also, if we have a dependency on cloud-init here, we wouldn't see any startup time improvements, would we?

mkulke commented:
> Currently, I'm adding the process-user-data service by default and changing the option based on whether cloud-config needs to be enabled or disabled. That seems to avoid using systemd overrides.

hmm, I don't think it's a good practice to make a tool skip itself based on a flag (`mytool -dontrunme`). I would rather have the caller decide whether it should invoke a binary or not (in our case the caller would be the kata-agent/agent-protocol-fowarder systemd units). Since we're already writing files into `/etc/defaults` conditionally, it wouldn't be a lot more complex to add systemd override files instead into `/etc/systemd/system/{kata-agent,protocol-forwarder-service}.service.d/foo.conf` or even just patching the process-user-data into an `ExecStart=true` noop (even though I think setting the proper unit dependencies would be cleaner).

But maybe that's a personal preference, others might want to chime in. @surajssd @kartikjoshi21 @jepio 

Overall, to me it looks like that the option to support both cloud-config and raw user-data adds a good chunk of complexity. I would opt for replacing the cloud-config option, at least initially. so we can test the user-data approach until we deem it solid and if we conclude we need to support both options (albeit cc_kbc will only work with process-user-data) we can add the toggle.

mkulke commented:
> I'm hitting the following error when using the same PR changes on AWS :-(
> 
> ```
> systemd[1]: multi-user.target: Found ordering cycle on kata-agent.service/start
> systemd[1]: multi-user.target: Found dependency on process-user-data.service/start
> systemd[1]: multi-user.target: Found dependency on cloud-init.target/start
> systemd[1]: multi-user.target: Found dependency on multi-user.target/start
> systemd[1]: multi-user.target: Job kata-agent.service/start deleted to break ordering cycle starting with multi-user.target/start
> systemd[1]: kata-agent.service: Found ordering cycle on process-user-data.service/start
> systemd[1]: kata-agent.service: Found dependency on cloud-init.target/start
> systemd[1]: kata-agent.service: Found dependency on multi-user.target/start
> systemd[1]: kata-agent.service: Found dependency on kata-agent.service/start
> systemd[1]: kata-agent.service: Job kata-agent.service/start deleted to break ordering cycle starting with kata-agent.service/start
> ```

It's a warning I think, but we need to resolve this. I am still able to spawn pods using this a podvm/caa daemonset built from this PR successfully (albeit not consistently, see the other PR):

```bash
journalctl -u process-user-data.service
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found ordering cycle on attestation-agent.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on kata-agent.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on process-user-data.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on cloud-init.target/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on cloud-final.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on multi-user.target/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Job attestation-agent.service/start deleted to break ordering cycle starting with multi-user.target/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found ordering cycle on kata-agent.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on process-user-data.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on cloud-init.target/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on cloud-final.service/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Found dependency on multi-user.target/start
Sep 06 08:15:45 localhost.localdomain systemd[1]: multi-user.target: Job kata-agent.service/start deleted to break ordering cycle starting with multi-user.target/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Found ordering cycle on process-user-data.service/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Found dependency on cloud-init.target/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Found dependency on cloud-final.service/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Found dependency on multi-user.target/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Found dependency on kata-agent.service/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: kata-agent.service: Job kata-agent.service/start deleted to break ordering cycle starting with kata-agent.service/start
Sep 06 08:15:53 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: Starting Process user data...
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: process-user-data version v0.7.0-dev
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:   commit: 7b9136c1e19953b92df96b1b82fd2964a0a9d00c-dirty
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:   go: go1.20.3
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: userDataUrl: http://169.254.169.254/metadata/instance/compute/userData?api-version=2021-01-01&format=text
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: Valid user data: {
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "pod-network": {
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "podip": "10.244.0.220/24",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "pod-hw-addr": "06:8f:1b:64:8e:57",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "interface": "eth0",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "worker-node-ip": "10.200.28.5/24",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "tunnel-type": "vxlan",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "routes": [
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:                   {
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:                       "Dst": "",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:                       "GW": "10.244.0.1",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:                       "Dev": "eth0"
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:                   }
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               ],
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "mtu": 1500,
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "index": 3,
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "vxlan-port": 8472,
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "vxlan-id": 555003,
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:               "dedicated": false
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           },
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "pod-namespace": "default",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "pod-name": "busybox-caa-78459759-qqdmh",
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "tls-server-key": "-----BEGIN PRIVATE KEY-----\nMIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgeZW6z158HCAxrvms\nJEJ0>
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "tls-server-cert": "-----BEGIN CERTIFICATE-----\nMIICBzCCAaygAwIBAgIQMtedJOaMPbOnaMMqPH6X1jAKBggqhkjOPQQDAjAjMSEw\nHwY>
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "tls-client-ca": "-----BEGIN CERTIFICATE-----\nMIIBbTCCARKgAwIBAgIRAJER0tT+VgRdX7u90kUiO9wwCgYIKoZIzj0EAwIwHDEa\nMBgGA>
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:           "aa-kbc-params": "cc_kbc::http://10.200.28.5:30080"
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]:       }
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: Wrote userData to file: /peerpod/daemon.json
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: Updated line: aa_kbc_params = "cc_kbc::http://10.200.28.5:30080"
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a process-user-data[854]: Updated agent config file: /etc/agent-config.toml
Sep 06 08:15:55 podvm-busybox-caa-78459759-qqdmh-4dbc441a systemd[1]: Finished Process user data.
```

bpradipt commented:
>Overall, to me it looks like that the option to support both cloud-config and raw user-data adds a good chunk of complexity. I would opt for replacing the cloud-config option, at least initially. so we can test the user-data approach until we deem it solid and if we conclude we need to support both options (albeit cc_kbc will only work with process-user-data) we can add the toggle.

It's indeed a bit complex. However replacing cloud-config completely might also break some stuff. I don't know for sure. Currently the two issues are related to systemd config. Let me see if I can fix them and based on that we can decide whether to switch completely to using metadata service for user data

bpradipt commented:
> hmm, I don't think it's a good practice to make a tool skip itself based on a flag (`mytool -dontrunme`). I would rather have the caller decide whether it should invoke a binary or not (in our case the caller would be the kata-agent/agent-protocol-fowarder systemd units). Since we're already writing files into `/etc/defaults` conditionally, it wouldn't be a lot more complex to add systemd override files instead into `/etc/systemd/system/{kata-agent,protocol-forwarder-service}.service.d/foo.conf` or even just patching the process-user-data into an `ExecStart=true` noop (even though I think setting the proper unit dependencies would be cleaner).

Actually my intention with adding `disable-cloud-config` was two folds - One was to eventually use process-user-data to even read from cloud-config and make any changes before kata-agent and agent-protocol-forwarder runs.
Second using the option to control what gets used to retrieve the user data. 

I do think the systemd override will still be needed to update the dependencies based on whether cloud-config or metadata service is used to fetch the user data.

Wdyt ?

bpradipt commented:
I tested the new changes on AWS successfully. So looks like nothing broke ;)
Now setting up cluster on Azure to test fetching of user data via cloud-config and metadata service.

mkulke commented:
> > hmm, I don't think it's a good practice to make a tool skip itself based on a flag (`mytool -dontrunme`). I would rather have the caller decide whether it should invoke a binary or not (in our case the caller would be the kata-agent/agent-protocol-fowarder systemd units). Since we're already writing files into `/etc/defaults` conditionally, it wouldn't be a lot more complex to add systemd override files instead into `/etc/systemd/system/{kata-agent,protocol-forwarder-service}.service.d/foo.conf` or even just patching the process-user-data into an `ExecStart=true` noop (even though I think setting the proper unit dependencies would be cleaner).
> 
> Actually my intention with adding `disable-cloud-config` was two folds - One was to eventually use process-user-data to even read from cloud-config and make any changes before kata-agent and agent-protocol-forwarder runs. Second using the option to control what gets used to retrieve the user data.
> 
> I do think the systemd override will still be needed to update the dependencies based on whether cloud-config or metadata service is used to fetch the user data.
> 
> Wdyt ?

hmm, cloud-config+cloud-init and the process-user-data tool would overlap in functionality as both cloud-init and the tool provision files to the podvm (see [here](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/pkg/util/cloudinit/cloudconfig.go#L46)), while the tool is also is able to make changes to the kata-agent configuration file. In fact the tool would be able to perform all tasks that we use cloud-config for atm.

If we want to keep the cloud-config option - which is fair - then I would suggest not to mix two approaches: the tool doesn't need to  know anything about cloud-config and it should only be called if cloud-init is not being used, which can be a build-time decision.

bpradipt commented:
> hmm, cloud-config+cloud-init and the process-user-data tool would overlap in functionality as both cloud-init and the tool provision files to the podvm (see [here](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/pkg/util/cloudinit/cloudconfig.go#L46)), while the tool is also is able to make changes to the kata-agent configuration file. In fact the tool would be able to perform all tasks that we use cloud-config for atm.

Right. Since with cloud-config, the files are already available inside the VM, my intention was to use the process-user-data tool to read those files and update the kata agent-config. Say making the kbc params available at /etc/kbc-params.json via cloud-config and process-user-data reading it and updating agent-config.  
When using cloud-config, the process-user-data adds a dependency to cloud-init.target via override to ensure it waits for the files to be available.

Do you still think we should use process-user-data only when fetching user-data using the metadata service ? Although the current process-user-data code doesn't have any support to read the local files (added via cloud-init+cloud-config) and update agent config. 
I can drop the changes and add process-user-data service only when not using cloud-config. Let me know.


A related update. I don't see the ordering issue with the latest changes. Can you kindly check.
And thanks a lot for reviewing and testing this PR.

mkulke commented:
> Right. Since with cloud-config, the files are already available inside the VM, my intention was to use the process-user-data tool to read those files and update the kata agent-config. Say making the kbc params available at /etc/kbc-params.json via cloud-config and process-user-data reading it and updating agent-config. When using cloud-config, the process-user-data adds a dependency to cloud-init.target via override to ensure it waits for the files to be available.
> 
> Do you still think we should use process-user-data only when fetching user-data using the metadata service ? Although the current process-user-data code doesn't have any support to read the local files (added via cloud-init+cloud-config) and update agent config. I can drop the changes and add process-user-data service only when not using cloud-config. Let me know.

Ok, putting it like this makes some sense indeed. What we could do:

- have a tool subcmd `process-user-data provision-files` to provision files from metadata service (into daemon.json and aa-kbc-params file), so it would behave like 
- have a tool subcmd  `process-user-data configure-agent` to set aa-kbc-params
 
If we disable cloud-config and need to provision the files manually we can createa systemd overlay w/ `ExecPreStart=`process-user-data provision-files`.
 
This way we'd add support for setting aa-kbc-params via cloud-config and have user-data as an option for a potential speedup.

mkulke commented:
> > A related update. I don't see the ordering issue with the latest changes. Can you kindly check. And thanks a lot for reviewing and testing this PR.
> 
> thanks for working on this, I'll check the next iteration as soon as I'll find the time.

I tested `f3b6d329e15401f77a823ca13146d1f459245f94` with 10+ pods, all came up. I did not test image decryption yet, but the aa_kbc string is properly replaced in the config.

bpradipt commented:
@mkulke I added subcommands for process-user-data and updated the systemd scripts.
Initial run on Azure was successful. I'll do some more tests. If you get time, ptal at the new changes

mkulke commented on podvm/files/etc/systemd/system/process-user-data.service:
```diff
@@ -0,0 +1,17 @@
+
+[Service]
+Type=oneshot
+EnvironmentFile=-/etc/default/process-user-data
```
That's probably a leftover?

bpradipt replied:
Ah yes..

mkulke commented on podvm/files/etc/systemd/system/process-user-data.service:
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Process user data
+After=network.target
+DefaultDependencies=no
```
probably a leftover too

bpradipt replied:
Intentional

mkulke commented on podvm/files/etc/systemd/system/process-user-data.service:
```diff
@@ -0,0 +1,17 @@
+[Service]
+Type=oneshot
+EnvironmentFile=-/etc/default/process-user-data
+ExecStart=/usr/local/bin/process-user-data provision-files
```
see comment on the kata-agent unit

```suggestion
ExecStartPre=/usr/local/bin/process-user-data provision-files
ExecStart=/usr/local/bin/process-user-data update-agent-config
```

mkulke commented on cmd/process-user-data/provision.go:
```diff
@@ -0,0 +1,249 @@
+		    "pod-name": "nginx-866fdb5bfb-b98nw",
+		    "tls-server-key": "-----BEGIN PRIVATE KEY-----\n....\n-----END PRIVATE KEY-----\n",
+		    "tls-server-cert": "-----BEGIN CERTIFICATE-----\n....\n-----END CERTIFICATE-----\n",
+		    "tls-client-ca": "-----BEGIN CERTIFICATE-----\n....\n-----END CERTIFICATE-----\n",
+			"aa-kbc-params": "cc_kbc::http://192.168.100.2:8080"
+
```
```suggestion
		    "tls-client-ca": "-----BEGIN CERTIFICATE-----\n....\n-----END CERTIFICATE-----\n",
		    "aa-kbc-params": "cc_kbc::http://192.168.100.2:8080"

```

mkulke reviewed (commented):
I'll test, but this looks good, apart from some nits, thanks.

bpradipt commented on podvm/files/etc/systemd/system/kata-agent.service:
```diff
@@ -1,11 +1,13 @@
 [Unit]
 Description=Kata Agent
 BindsTo=netns@podns.service
+Wants=process-user-data.service
 After=network.target netns@podns.service
 
 [Service]
 ExecStart=/usr/local/bin/kata-agent --config /etc/agent-config.toml
 ExecStartPre=-umount /sys/fs/cgroup/misc
+ExecStartPre=/usr/local/bin/process-user-data update-agent-config
```
Adding process-user-data.service in "After" created the cyclic dependency loop error earlier. I can retry

bpradipt commented on podvm/qcow2/misc-settings.sh:
```diff
@@ -83,6 +83,23 @@ TLS_OPTIONS=-cert-file /etc/certificates/tls.crt -cert-key /etc/certificates/tls
+    After=cloud-init.target
+
+    [Service]
+    ExecStart=
+    ExecStart=true
```
Yes

bpradipt replied:
kata-agent had the dependency on cloud-init, which I move to process-user-data. If cloud-config is not used, process-user-data provisions the file. And if cloud-config is used, process-user-data returns true but adds the cloud-init dep.
Also processing the user-data to update the agent config is useful for both the cases hence it's done before starting kata-agent.

mkulke replied:
yup the agent config should happen in each case. it just itches me a bit that we have to noop a unit file, instead of skipping it. that probably indicates that we are doing something wrong or in a weird way. 

Still, it works and it's just a nitpick. We can look into that in a separate PR.

bpradipt replied:
I'll try your suggestion on both Azure and AWS tomorrow. And if this works then we won't have to use noop.

bpradipt replied:
With minor modification to remove ordering issue, I was able to use your suggestion and got it successfully working for AWS as well. Thanks

mkulke reviewed (approved):
Tested `d58340e` (fresh PodVM + CAA image build) successfully w/ cloud-config & user-data. Both work and aa_kbc_param is properly configured in the latter case. Great work, thanks again.

Some perf numbers, kata-agent starts immediately after the VM has been successfully created.

```
2023/09/07 18:35:59 [adaptor/cloud/azure] CreateInstance: name: "podvm-busybox-caa-5f99bd66c7-pj5vv-79a793af"
2023/09/07 18:36:46 [adaptor/cloud/azure] created VM successfully: /subscriptions/REDACTED/resourceGroups/mgns/providers/Microsoft.Compute/virtualMachine
...
2023/09/07 18:36:46 [adaptor/proxy] CreateSandbox: hostname:busybox-caa-5f99bd66c7-pj5vv sandboxId:79a793afae37f08d89981396d6d05af4266704db2d091301608bc14ef04ba80a
...
2023/09/07 18:36:49 [adaptor/proxy] StartContainer: containerID:b37f200743b559b26c63971c7f3829dc6a26e3dab857a34ac7b0a202792b63ad
```

bpradipt commented:
@kartikjoshi21 @mkulke is it possible to run an e2e ci on Azure for this PR ?

mkulke commented:
> @kartikjoshi21 @mkulke is it possible to run an e2e ci on Azure for this PR ?

Sadly this is not possible with forks and it's also not trivial to run them manually.

bpradipt commented:
> > @kartikjoshi21 @mkulke is it possible to run an e2e ci on Azure for this PR ?
> 
> Sadly this is not possible with forks and it's also not trivial to run them manually.

Np..

snir911 commented on pkg/adaptor/cloud/azure/provider.go:
```diff
@@ -149,15 +149,40 @@ func (p *azureProvider) createNetworkInterface(ctx context.Context, nicName stri
+		//Convert cloudConfigData to base64
+		b64EncData = base64.StdEncoding.EncodeToString([]byte(cloudConfigData))
+	} else {
+		userData := strings.Split(cloudConfigData, "content: |")[1]
```
Maybe i missed that, is there size limit for userdata passed using imds? is it validated? IIRC data size is checked, might be different/irrelevant now

bpradipt replied:
There is indeed a limit to userData - 64KB. I think it's better to check the size and throw error if the size is more than 64KB.
I'll add the check. Thanks @snir911

bpradipt replied:
I have added the explicit check for Azure. As per the Azure doc, both user data (via metadata service) or custom data (via config drive) cannot exceed 64KB. This check was not present in the original code as well. So I added it as a separate commit.

snir911 commented on Makefile:
```diff
@@ -12,7 +12,7 @@ RELEASE_BUILD ?= false
 GOOPTIONS   ?= GOOS=linux GOARCH=$(ARCH) CGO_ENABLED=0
 GOFLAGS     ?=
-BINARIES    := cloud-api-adaptor agent-protocol-forwarder
+BINARIES    := cloud-api-adaptor agent-protocol-forwarder process-user-data
```
Shouldn't it be added for Azure only?

bpradipt replied:
@snir911 this is a generic command addition which includes two sub commands. One  to process user-data to update the agent config and another to fetch the metadata service from metadata service.

The default mechanism to fetch the user-data is still via config-drive.  And if metadata fetching is needed then the image needs to be built with DISABLE_CLOUD_CONFIG=true. 
For updating the agent-config using the user-data from the file is common for all providers.

snir911 reviewed (approved):
Thanks!
